src/main/scala/org/apache/spark/sql/execution/datasources/greenplum/GreenplumUtils.scala

```diff
@@ -60,71 +61,157 @@ object GreenplumUtils extends Logging {
    *
+   * Copy data to greenplum and append to relative gptable.
+   *
    * @param df the [[DataFrame]] will be copy to the Greenplum
    * @param schema the table schema in Greemnplum
    * @param options Options for the Greenplum data source
    */
-  def copyToGreenplum(df: DataFrame, schema: StructType, options: GreenplumOptions): Unit = {
-    val valueConverters: Array[(Row, Int) => String] =
-      schema.map(s => makeConverter(s.dataType, options)).toArray
+  def copyAppendToGreenplum(
+      df: DataFrame,
+      schema: StructType,
+      options: GreenplumOptions): Unit = {
+    df.foreachPartition { rows =>
+      copyParition(rows, options, schema, options.table)
+    }
+  }
 
-    def convertRow(row: Row): Array[Byte] = {
-      var i = 0
-      val values = new Array[String](schema.length)
-      while (i < schema.length) {
-        if (!row.isNullAt(i)) {
-          values(i) = convertValue(valueConverters(i).apply(row, i))
-        } else {
-          values(i) = "NULL"
-        }
-        i += 1
-      }
-      (values.mkString(options.delimiter) + "\n").getBytes("UTF-8")
+  /**
+   * https://www.postgresql.org/docs/9.2/sql-copy.html
+   *
+   * Copy data to greenplum and overwrite relative gptable,
+   * which can be dropped or does not exist.
+   *
+   * @param df the [[DataFrame]] will be copy to the Greenplum
+   * @param schema the table schema in Greemnplum
+   * @param options Options for the Greenplum data source
+   */
+  def copyOverwriteToGreenplum(
+      df: DataFrame,
+      schema: StructType,
+      options: GreenplumOptions): Unit = {
+    val randomString = UUID.randomUUID().toString.flatMap {
+      case '-' => ""
+      case c => s"$c"
     }
+    val tempTable = s"sparkGpTmp$randomString"
+    val strSchema = JdbcUtils.schemaString(df, options.url, options.createTableColumnTypes)
+    val createTempTbl = s"CREATE TABLE $tempTable ($strSchema) ${options.createTableOptions}"
+
+    val conn = JdbcUtils.createConnectionFactory(options)()
+
+    try {
+      executeStatement(conn, createTempTbl)
+      val accumulator = df.sparkSession.sparkContext.longAccumulator("copySuccess")
+      val partNum = df.rdd.getNumPartitions
+
+      df.foreachPartition { rows =>
+        copyParition(rows, options, schema, tempTable, accumulator)
+      }
+
+      if (accumulator.value == partNum) {
+        val dropTbl = s"DROP TABLE IF EXISTS ${options.table}"
+        executeStatement(conn, dropTbl)
 
-    def convertValue(str: String): String = {
-      assert(options.delimiter.length == 1, "The delimiter should be a single character.")
-      val delimiter = options.delimiter.charAt(0)
-      str.flatMap {
-        case '\\' => "\\\\"
-        case '\n' => "\\n"
-        case '\r' => "\\r"
-        case `delimiter` => s"\\$delimiter"
-        case c => s"$c"
+        val renameTempTbl = s"ALTER TABLE $tempTable RENAME TO ${options.table}"
+        executeStatement(conn, renameTempTbl)
+      } else {
```

Comment:
why do we fail silently here

Reply:
If there are some errors occured, relative sql statement will throw exception.
So should we throw exception for just closing connection failed?

Reply:
Not for capturing those exceptions. If we have task failures, we should throw an exception here. do not just delete tmp table here. Otherwise, the user will know nothing whether it success or not.

Reply:
> Not for capturing those exceptions. If we have task failures, we should throw an exception here. do not just delete tmp table here. Otherwise, the user will know nothing whether it success or not.

Ok.